File: _python_backup/services/ai_service.py
"""
AI interpretation service
"""

from typing import Dict, Optional
from domain.models import AnalysisResults
from domain.exceptions import VNGError
from modules.ai_interpreter import get_interpretation as _get_interpretation
from modules.ai_interpreter import build_interpretation_prompt
from config.settings import settings


class AIService:
    """Service for AI-powered interpretation"""
    
    @staticmethod
    def get_interpretation(analysis_results: AnalysisResults) -> Optional[str]:
        """
        Get AI interpretation of analysis results
        
        Args:
            analysis_results: AnalysisResults domain model
            
        Returns:
            Interpretation text or None if error
            
        Raises:
            VNGError: If API key is missing or request fails
        """
        api_key = settings.get_api_key()
        if not api_key:
            raise VNGError("AI API key not configured")
        
        # Convert domain model to dict format for legacy function
        results_dict = {
            category: {
                metric: {
                    'values': data.values,
                    'flags': data.flags,
                    'delta': data.delta,
                    'percent_change': data.percent_change,
                    'std_dev': data.std_dev
                }
                for metric, data in result.metrics.items()
            }
            for category, result in analysis_results.results.items()
        }
        
        try:
            return _get_interpretation(
                api_key,
                results_dict,
                analysis_results.file_count
            )
        except Exception as e:
            raise VNGError(f"Failed to get AI interpretation: {str(e)}") from e

File: js/src/services/aiService.js
/**
 * AI service for interpretation - JavaScript version
 * Note: This is a placeholder implementation that would require backend API integration
 */

const { VNGError } = typeof module !== 'undefined' && module.exports
    ? require('../domain/exceptions')
    : window.VNGExceptions;

/**
 * Service for AI-powered interpretation
 * Currently a placeholder - requires backend API integration
 */
class AIService {
    /**
     * Generate AI interpretation of analysis results
     * @param {Object} analysisResults - Analysis results dictionary
     * @param {Array} fileDataList - List of file data
     * @returns {Promise<string>} AI interpretation text
     */
    static async generateInterpretation(analysisResults, fileDataList) {
        // This is a placeholder implementation
        // In a real implementation, this would make API calls to an AI service

        return new Promise((resolve, reject) => {
            // Simulate API delay
            setTimeout(() => {
                try {
                    const interpretation = this._generateMockInterpretation(analysisResults, fileDataList);
                    resolve(interpretation);
                } catch (error) {
                    reject(new VNGError(`AI interpretation failed: ${error.message}`));
                }
            }, 1000);
        });
    }

    /**
     * Generate mock interpretation for demonstration
     * @private
     * @param {Object} analysisResults - Analysis results
     * @param {Array} fileDataList - File data list
     * @returns {string} Mock interpretation text
     */
    static _generateMockInterpretation(analysisResults, fileDataList) {
        let interpretation = `# VNG Analysis Clinical Interpretation\n\n`;

        interpretation += `## Overview\n`;
        interpretation += `Analysis of ${fileDataList.length} VNG report(s) completed.\n\n`;

        // Count flagged metrics
        let totalFlagged = 0;
        let significantChanges = 0;

        for (const category in analysisResults) {
            for (const metric in analysisResults[category]) {
                const data = analysisResults[category][metric];
                if (data.flags && data.flags.some(flag => flag)) {
                    totalFlagged++;
                }
                if (data.percentChange && Math.abs(data.percentChange) > 10) {
                    significantChanges++;
                }
            }
        }

        interpretation += `## Key Findings\n`;
        interpretation += `- **Files Analyzed**: ${fileDataList.length}\n`;
        interpretation += `- **Flagged Metrics**: ${totalFlagged}\n`;
        interpretation += `- **Significant Changes (>10%)**: ${significantChanges}\n\n`;

        interpretation += `## Clinical Considerations\n\n`;

        if (totalFlagged > 0) {
            interpretation += `⚠️ **Abnormal Findings Detected**: ${totalFlagged} metrics were flagged as outside normal ranges. `;
            interpretation += `Please review the flagged items in the detailed analysis tab for specific clinical implications.\n\n`;
        }

        if (significantChanges > 0) {
            interpretation += `📈 **Significant Changes Observed**: ${significantChanges} metrics showed changes greater than 10% between reports. `;
            interpretation += `This may indicate progression or improvement in vestibular function.\n\n`;
        }

        interpretation += `## Recommendations\n`;
        interpretation += `1. **Review Flagged Metrics**: Pay special attention to any metrics marked with warning flags.\n`;
        interpretation += `2. **Clinical Correlation**: Correlate these findings with patient symptoms and clinical presentation.\n`;
        interpretation += `3. **Follow-up Testing**: Consider repeat testing if significant changes are observed.\n`;
        interpretation += `4. **Documentation**: Ensure all findings are properly documented in the patient record.\n\n`;

        interpretation += `---\n*This interpretation is generated for demonstration purposes. `;
        interpretation += `For actual clinical use, please consult with a qualified healthcare professional.*\n`;

        return interpretation;
    }

    /**
     * Check if AI service is available
     * @returns {boolean} True if service is available
     */
    static isAvailable() {
        // In a real implementation, this would check API connectivity
        return false; // Currently returns false as this is a placeholder
    }

    /**
     * Get service status information
     * @returns {Object} Status information
     */
    static getStatus() {
        return {
            available: this.isAvailable(),
            service: 'AI Interpretation Service',
            status: 'Placeholder Implementation',
            message: 'Backend API integration required for full functionality'
        };
    }
}

// Export service
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { AIService };
} else if (typeof window !== 'undefined') {
    window.VNGServices = window.VNGServices || {};
    window.VNGServices.AIService = AIService;
}
